### PR TITLE
Using YAML instead of INI files.

### PR DESCRIPTION
**Pull request by facocalj**
Made the changes for using YAML configuration files instead of INI, which enables multi line hooks

Apparently the ConfigParser was reading the system-wide config file and updating it with the user's hooks and such. This functionality is maintained by updating a dictionary with the user's values.

### CODE REVIEW

**Comment by edio** [comment]
Thanks a lot for the PR. I'll take a look at this later today.

In the meantime, do you think it's worth supporting both ini and yaml together with a warning, saying that ini is deprecated?

**Comment by facocalj** [comment]
Initially I thought that keeping both of them would be a good idea, specially that it wouldn't break other user's setups on update (even if migrating their hooks would be dead easy).

But what if they have both YAML and INI files? Should we favour the new format? 
What about the new functionality (multi line hooks), that is not present using INI files? We would need to emphasise the difference in the documentation, or at least throw a proper error regarding the line break

**Comment by edio** [comment]
@facocalj I'd absolutely made randrctl to prefer yaml instead of INI.

If users keep using INI, warning should be printed, but feature would keep working the way before the update.

If users have yaml, they will have new features that come with it, including multiline values.

Feel free to update documentation as well. 
Or I could do this after your PR: that README begs for cleanup anyway.

File: randrctl/ctl.py [hunk 1]
```diff
@@ -201,12 +201,21 @@ def _configs(self):
...
-
         config_files = list(self._configs())
         config_files.reverse()  # reverse, so config file from the preferred home comes last (has highest precedence)
-        read = config.read(config_files)
-        logger.debug("read configuration from %s", read)
+
+        # Empty container to hold the configuration
+        config = {}
+        # Iterate over the list of config files
+        for config_file in config_files:
```
Comment (edio) [inline randrctl/ctl.py]: Although `ConfigParser` had `strict=False` and parsed 2 configs, I think any sort of merge is counterintuitive. It should be either system-wide config (which should be deprecated eventually I believe), or user config. Never both.

Behavior is even harmful. Imagine, system config defines prior and post hooks to STOP and CONT i3, then user defines post hook to do something else, so system hook becomes overridden by users config and i3 receives SIG_STOP but never SIG_CONT.

Do you agree, that having only one config at a time would be simpler, cleaner and just better?

If yes, would you be willing to fix this part in your PR :) ?

Comment (facocalj) [reply randrctl/ctl.py]: I was just trying to mimic how ConfigParser was being used, but I totally agree with your reasoning

In short, randrctl should load the system-wide config, but prefer the user's config (if available) 
And favour YAML in both cases, outputting warning messages if using INI. 
Did I get it right? If so, I can make the changes either today or tomorrow

Comment (edio) [reply randrctl/ctl.py]: Yes, this is correct (just to clarify: system config may be also not available).

Looking forward for the changes! Thank you.

File: randrctl/ctl.py [hunk 2]
```diff
@@ -193,29 +194,80 @@ def _config_safe(self, config: dict, section: str, property: str):
...
+          try:
+            read = config_parser.read(ini_files)
+          except Exception as e:
+            logger.error("error reading configuration files")
+          else:
+            config = {section: dict(config_parser.items(section)) for section in config_parser.sections()}
+
+        # Otherwise, there are no configuration files whatsoever
+        else:
+          logger.error("there are no configuration files available")
```
Comment (edio) [inline randrctl/ctl.py]: Maybe from this method point of view it is an error, but from application point of view it is not — _randrctl_ should work fine without configs present. We should not log this as error. As I proposed above, let's just log the name of the config file that we picked among all available

File: randrctl/ctl.py [hunk 3]
```diff
@@ -193,29 +194,80 @@ def _config_safe(self, config: dict, section: str, property: str):
...
+        :return: dictionary of config files in defined home directories if exist
         """
+        # Empty container to hold the configs
+        configs = {}
         for homedir in self.homes:
-            config = os.path.join(homedir, self.config_name)
-            if os.path.isfile(config):
-                yield config
+            # Configs will be either user's (prefered) or system-wide (other)
+            is_favourite = 'preferred' if homedir == self.preferred_home else 'other'
```
Comment (edio) [inline randrctl/ctl.py]: I think `is_favourite` is not exactly is favourite but generally denotes the group in which config falls in. Should be renamed to something that represents its meaning better. If you do not have better ideas, what about `config_group`?

Comment (edio) [inline randrctl/ctl.py]: extract `'preferred'` and `'other'` string literals to constants

File: randrctl/ctl.py [hunk 4]
```diff
@@ -193,29 +194,80 @@ def _config_safe(self, config: dict, section: str, property: str):
...
+        config_files = self._configs()
 
-    def get_randrctl(self):
-        config_files = list(self._configs())
-        config_files.reverse()  # reverse, so config file from the preferred home comes last (has highest precedence)
+        # If available, use preferred configs, otherwise use the others
+        use = 'preferred' if config_files.get('preferred') else 'other'
+
+        if use == 'other':
+          logger.warning("using system-wide configuration files")
```
Comment (edio) [inline randrctl/ctl.py]: I wouldn't log this as warning. How about just logging as debug the name of the config that was picked eventually?

File: randrctl/ctl.py [hunk 5]
```diff
@@ -193,29 +194,80 @@ def _config_safe(self, config: dict, section: str, property: str):
...
+          logger.warning("loading INI configuration files")
+
+          # Instantiate config parser
+          config_parser = ConfigParser(allow_no_value=True, strict=False)
+
+          # Try to read the config files
+          try:
+            read = config_parser.read(ini_files)
+          except Exception as e:
+            logger.error("error reading configuration files")
```
Comment (edio) [inline randrctl/ctl.py]: warning

File: randrctl/ctl.py [hunk 6]
```diff
@@ -193,29 +194,80 @@ def _config_safe(self, config: dict, section: str, property: str):
...
+              # Try to parse the YAML file  and update the configuration dictionary
+              try:
+                config.update(load(stream))
+              except YAMLError as e:
+                logger.error("error reading configuration file %s", config_file)
+              else:
+                logger.debug("read configuration from %s", config_file)
+
+        elif ini_files:
+          logger.warning("loading INI configuration files")
```
Comment (edio) [inline randrctl/ctl.py]: Message should say: 
> INI configuration is deprecated and will be removed in next minor release
> Please convert &lt;filepath&gt; to yaml format
> Visit https://github.com/edio/randrctl#priorpost-hooks for details

File: randrctl/ctl.py [hunk 7]
```diff
@@ -193,29 +194,80 @@ def _config_safe(self, config: dict, section: str, property: str):
...
-                    config.update(yaml.load(stream))
-                except yaml.YAMLError as e:
-                    logger.error("error reading configuration file %s", config_file)
-                else:
-                    logger.debug("read configuration from %s", config_file)
+              # Try to parse the YAML file  and update the configuration dictionary
+              try:
+                config.update(load(stream))
+              except YAMLError as e:
+                logger.error("error reading configuration file %s", config_file)
```
Comment (edio) [inline randrctl/ctl.py]: warning

**Comment by facocalj** [comment]
Hi @edio,

I've made all the changes on 684e64f5f3653609f1b236f047f314bc8a567228. Let me know what you think about them

File: randrctl/ctl.py [hunk 8]
```diff
@@ -181,32 +185,92 @@ def ensure_homes(self):
...
+            try:
+              config.update(load(stream))
+            except YAMLError as e:
+              logger.warning("error reading configuration file %s", yaml_files[0])
+            else:
+              logger.debug("read configuration from %s", yaml_files[0])
+
+        elif ini_files:
+          # Warning message indicating deprecation
+          logger.warning("INI configuration is deprecated and will be removed in next minor release")
```
Comment (edio) [inline randrctl/ctl.py]: I think, we'd better log this as a single multi-line message. But otherwise lgtm.

Comment (facocalj) [reply randrctl/ctl.py]: Done!

**Comment by edio** [comment]
Thanks for your contribution!

**Pull request merged**